connman/plugins/sailfish_ofono.c

```diff
@@ -137,6 +138,121 @@ struct plugin_data {
 	gulong manager_handler_id[MANAGER_HANDLER_COUNT];
 };
 
+struct rtnl_gw {
+	int index;
+	char *dst;
+	char *gw;
+	int metric;
+};
+
+static GSList *rtnl_gw_list = NULL;
+
+static void free_rtnl_gw(gpointer data)
+{
+	struct rtnl_gw *item = data;
+
+	g_free(item->dst);
+	g_free(item->gw);
+	g_free(item);
+}
+
+static gint route_compare(gconstpointer a, gconstpointer b)
+{
+	const struct rtnl_gw *route_a = a;
+	const struct rtnl_gw *route_b = b;
+
+	if(route_a->index == route_b->index &&
+				!g_strcmp0(route_a->dst, route_b->dst) &&
+				!g_strcmp0(route_a->gw, route_b->gw) &&
+				route_a->metric == route_b->metric)
+		return 0;
+
+	return -1;
+}
+
+static void add_gateway_to_rtnl_list(int index, const char *dst,
+						const char *gw, int metric)
+{
+	struct rtnl_gw *item;
+
+	DBG("%d %s %s %d", index, dst, gw, metric);
+
+	item = g_try_new0(struct rtnl_gw, 1);
+	if (!item)
+		return;
+
+	item->index = index;
+	item->dst = g_strdup(dst);
+	item->gw = g_strdup(gw);
+	item->metric = metric;
+
+	if (g_slist_find_custom(rtnl_gw_list, item, route_compare)) {
+		free_rtnl_gw(item);
+		DBG("already in list");
+		return;
+	}
+
+	rtnl_gw_list = g_slist_prepend(rtnl_gw_list, item);
+	DBG("added");
+}
+
+static void del_gateway_from_rtnl_list(int index, const char *dst,
+						const char *gw, int metric)
+{
+	struct rtnl_gw *item;
+	GSList *iter;
+
+	DBG("%d %s %s %d", index, dst, gw, metric);
+
+	if (!rtnl_gw_list)
+		return;
+
+	for (iter = rtnl_gw_list; iter; iter = iter->next) {
+		item = iter->data;
+		if(item->index == index && !g_strcmp0(item->dst, dst) &&
+						!g_strcmp0(item->gw, gw) &&
+						item->metric == metric) {
+			rtnl_gw_list = g_slist_remove_link(rtnl_gw_list, iter);
+			free_rtnl_gw(item);
+			g_slist_free_1(iter);
+
+			DBG("deleted");
+			return;
+		}
+	}
+}
+
+static void handle_rtnl_gw(gpointer data, gpointer user_data)
+{
+	struct rtnl_gw *item = data;
+	int index = GPOINTER_TO_INT(user_data);
+	int err;
+
+	/* Delete all gateways that are not for the current interface */
+	if (item->index == index)
+		return;
+
+	DBG("deleting %d %s %s %d", item->index, item->dst, item->gw,
+								item->metric);
+
+	/* Interested only in default gw routes that have no prefix */
```

Comment:
Just a reminder for self `struct in6_rtmsg` does contain the prefix length as well.

Reply:
This was available otherwise.